### PR TITLE
Changed blinded block storage to store payloads indexed on block root.

### PR DESCRIPTION
**Pull request by rolfyone**
By using block root we'll be able to more efficiently prune once we're not storing payloads for finalized data.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

## Documentation

- [X] I thought about documentation and added the `doc-change-required` label to this PR if updates are required.

## Changelog

- [X] I thought about adding a changelog entry, and added one if I deemed necessary.

### CODE REVIEW

File: storage/src/main/java/tech/pegasys/teku/storage/server/kvstore/BlindedBlockKvStoreDatabase.java [hunk 1]
```diff
@@ -80,37 +81,50 @@ protected FinalizedUpdaterBlinded finalizedUpdater() {
...
       return maybeBlock;
     }
     final SignedBeaconBlock block = maybeBlock.get();
-    return Optional.of(getUnblindedBlock(block));
+    return Optional.of(getUnblindedBlock(blockRoot, block));
   }
 
-  private SignedBeaconBlock getUnblindedBlock(final SignedBeaconBlock block) {
+  private SignedBeaconBlock getUnblindedBlock(
+      final Bytes32 blockRoot, final SignedBeaconBlock block) {
```
Comment (ajsutton) [inline storage/src/main/java/tech/pegasys/teku/storage/server/kvstore/BlindedBlockKvStoreDatabase.java]: nit: I'd probably swap these arguments around so we're consistently using `block, blockRoot`.

File: storage/src/main/java/tech/pegasys/teku/storage/server/kvstore/BlindedBlockKvStoreDatabase.java [hunk 2]
```diff
@@ -80,37 +81,50 @@ protected FinalizedUpdaterBlinded finalizedUpdater() {
...
+          spec.atSlot(block.getSlot()).getSchemaDefinitions(),
+          SchemaDefinitionsBellatrix.required(spec.atSlot(block.getSlot()).getSchemaDefinitions())
+              .getExecutionPayloadSchema()
+              .getDefault());
+    }
+
+    final Optional<Bytes> maybePayload = dao.getExecutionPayload(blockRoot);
     if (maybePayload.isEmpty()) {
-      return block;
+      //      return block;
```
Comment (ajsutton) [inline storage/src/main/java/tech/pegasys/teku/storage/server/kvstore/BlindedBlockKvStoreDatabase.java]: nit: remove comment.

File: storage/src/main/java/tech/pegasys/teku/storage/server/kvstore/BlindedBlockKvStoreDatabase.java [hunk 3]
```diff
@@ -172,7 +186,12 @@ protected void storeFinalizedBlocksToDao(final Collection<SignedBeaconBlock> blo
                 .getMessage()
                 .getBody()
                 .getOptionalExecutionPayload()
-                .ifPresent(updater::addExecutionPayload);
+                .ifPresent(
+                    payload -> {
+                      if (!payload.isDefault()) {
+                        updater.addExecutionPayload(block.getRoot(), payload);
+                      }
+                    });
```
Comment (ajsutton) [inline storage/src/main/java/tech/pegasys/teku/storage/server/kvstore/BlindedBlockKvStoreDatabase.java]: Could use:
```suggestion
                .filter(payload -> !payload.isDefault)
                .ifPresent(updater::addExecutionPayload));
```

**Review by ajsutton — APPROVED** [review]
LGTM.

File: storage/src/main/java/tech/pegasys/teku/storage/server/kvstore/BlindedBlockKvStoreDatabase.java (hunk 1 above)
Comment (rolfyone) [reply storage/src/main/java/tech/pegasys/teku/storage/server/kvstore/BlindedBlockKvStoreDatabase.java]: usually we put index first so i changed it to this order...

File: storage/src/main/java/tech/pegasys/teku/storage/server/kvstore/BlindedBlockKvStoreDatabase.java (hunk 3 above)
Comment (rolfyone) [reply storage/src/main/java/tech/pegasys/teku/storage/server/kvstore/BlindedBlockKvStoreDatabase.java]: thats neat, i like it

File: storage/src/main/java/tech/pegasys/teku/storage/server/kvstore/BlindedBlockKvStoreDatabase.java (hunk 2 above)
Comment (rolfyone) [reply storage/src/main/java/tech/pegasys/teku/storage/server/kvstore/BlindedBlockKvStoreDatabase.java]: my bad, sorry about that.

File: storage/src/main/java/tech/pegasys/teku/storage/server/kvstore/BlindedBlockKvStoreDatabase.java (hunk 1 above)
Comment (ajsutton) [reply storage/src/main/java/tech/pegasys/teku/storage/server/kvstore/BlindedBlockKvStoreDatabase.java]: Either way is fine, I'd just make the two `getUnblindedBlock` methods have it the same way round.

**Pull request merged**